[Docs] Add Torch.Export Guides

YifanShenSZ opened the pull request:
This PR adds the doc guides for torch.export conversion support
1. How to export & convert a torch model
2. OpenELM as an e2e example

Corresponding HTML update: https://github.com/apple/coremltools/pull/2357

CI
~~✅ https://gitlab.com/coremltools1/coremltools/-/commit/7de44352f271d9a80b1da80e3f13d4f6d47e26fc/pipelines~~
✅ https://gitlab.com/coremltools1/coremltools/-/commit/ed27e93390c6cc2b3f3ed54de386fd282cd21235/pipelines

TobyRoseman commented on docs-guides/source/convert-llm.md:
```diff
@@ -0,0 +1,151 @@
+# Converting a Large Language Model
+
+The following example shows how to convert into Core ML an [OpenELM](https://huggingface.co/apple/OpenELM) model trained using PyTorch. OpenELM is a family of Open Efficient Language Models. The small size variants are suitable for mobile and embedded language applications.
```
This sentence is a bit awkward. I would rewrite: The following example shows how to convert a PyTorch OpenELM model to Core ML.

YifanShenSZ replied:
Revised as suggested

TobyRoseman commented on docs-guides/source/convert-llm.md:
```diff
@@ -0,0 +1,151 @@
+# prompt
+input_ids = tokenized_prompt.unsqueeze(0)
+logits = torch_model(input_ids)[0]
+output_id = torch.argmax(logits, -1)[:, -1 :]
```
I'm having a difficult time understanding this line and the for-loop that follows. Please make the code more readable, better variable names and comment.

YifanShenSZ replied:
Added more comments as suggested

TobyRoseman commented on docs-guides/source/convert-llm.md:
```diff
@@ -0,0 +1,151 @@
+tokenized_prompt = np.int32(tokenized_prompt.detach().numpy())
+input_ids = np.expand_dims(tokenized_prompt, axis=0)
+logits = list(mlmodel.predict({"input_ids": input_ids}).values())[0]
+output_id = np.argmax(logits, -1)[:, -1 :]
```
I'm also not really understanding this line or the following for-loop. Please try to make it more readable.

YifanShenSZ replied:
Same here

TobyRoseman commented on docs-guides/source/convert-llm.md:
```diff
@@ -0,0 +1,151 @@
+logits = list(mlmodel.predict({"input_ids": input_ids}).values())[0]
+output_id = np.argmax(logits, -1)[:, -1 :]
+# extend
+for i in range(64):
```
Rather than having `64` repeated twice, consider creating a variable. That let you give it a meaningful name which would help with readability.

YifanShenSZ replied:
Replaced `64` with `max_sequence_length`

TobyRoseman commented on docs-guides/source/model-exporting.md:
````diff
@@ -0,0 +1,111 @@
+
+```{admonition} Exporting Limitations
+
+Torch.export has limitations: graph breaks, data/shape-dependent control flow, and missing fake/meta/abtract kernels for operators. See [limitations of torch.export](https://pytorch.org/docs/stable/export.html#limitations-of-torch-export)
````
I don't think we need to talk much about torch.export limitations in this doc. Also this information may become incorrect over time. I would just include that link and leave it a that.

YifanShenSZ replied:
Revised as suggested

aseemw commented:
The entry point to the topic of PyTorch conversion is [this page](https://apple.github.io/coremltools/docs-guides/source/convert-pytorch-workflow.html ). 

Hence we need to update this page, by adding a section up top that clearly explains the following points (I'm stating roughly what it should cover below)

(I think we can update this page now, to act as a summary of all the formats supported and what to use when. We can delete the example that it currently has and let that be covered in the trace page)

- there are 3 ways to convert a torch model 
      
    - one that uses `torch.jit.script` . This is the method that's stable and has been supported since coremltools 4.0. This is the recommended flow for the most stable experience, as of now. Later will be migrated to torch.export as that becomes the officially recommended way to capture torch graphs. Point to the model tracing page, for details and examples ([this page](https://apple.github.io/coremltools/docs-guides/source/pytorch-conversion-examples.html), [this page](https://apple.github.io/coremltools/docs-guides/source/convert-a-torchvision-model-from-pytorch.html)). 
    
    - one that uses `torch.jit.script`. This was never supported in an official way,  it was always experimental and it recommended to only use on specific cases, when there is control flow around a model that can be traced.  Point to the page that discusses this topic and the example (this [page](https://apple.github.io/coremltools/docs-guides/source/convert-nlp-model.html))
   
     - the export path. This support is newly added in 8.0, currently in beta state, in line with the export API status in PyTorch. As of 8.0, so and so models can be converted, total op test coverage is roughly ~60%. Recommend to use this path, with the knowledge that it is in active development. Point to the page on export and example (the open elm one).

aseemw reviewed (changes requested):
The entry point to the topic of PyTorch conversion is [this page](https://apple.github.io/coremltools/docs-guides/source/convert-pytorch-workflow.html ). 

Hence we need to update this page, by adding a section up top that clearly explains the following points (I'm stating roughly what it should cover below)

(I think we can update this page now, to act as a summary of all the formats supported and what to use when. We can delete the example that it currently has and let that be covered in the trace page)

- there are 3 ways to convert a torch model 
      
    - one that uses `torch.jit.script` . This is the method that's stable and has been supported since coremltools 4.0. This is the recommended flow for the most stable experience, as of now. Later will be migrated to torch.export as that becomes the officially recommended way to capture torch graphs. Point to the model tracing page, for details and examples ([this page](https://apple.github.io/coremltools/docs-guides/source/pytorch-conversion-examples.html), [this page](https://apple.github.io/coremltools/docs-guides/source/convert-a-torchvision-model-from-pytorch.html)). 
    
    - one that uses `torch.jit.script`. This was never supported in an official way,  it was always experimental and it recommended to only use on specific cases, when there is control flow around a model that can be traced.  Point to the page that discusses this topic and the example (this [page](https://apple.github.io/coremltools/docs-guides/source/convert-nlp-model.html))
   
     - the export path. This support is newly added in 8.0, currently in beta state, in line with the export API status in PyTorch. As of 8.0, so and so models can be converted, total op test coverage is roughly ~60%. Recommend to use this path, with the knowledge that it is in active development. Point to the page on export and example (the open elm one).

aseemw commented on docs-guides/source/model-exporting.md:
```diff
@@ -0,0 +1,111 @@
+
+# Model Exporting
+
+The easiest way to generate ExportedProgram for your model is to use PyTorch's [torch.export.export](https://pytorch.org/docs/stable/export.html#torch.export.export). Exporting runs an example input tensor through your model, and captures the operations that are invoked as that input makes its way through the model's layers.
```
"easiest" because? Do we mean to say, the recommended way?

aseemw commented on docs-guides/source/model-exporting.md:
```diff
@@ -0,0 +1,111 @@
+## Difference from Tracing
+For tracing, `ct.convert` requires the `inputs` arg from user. This is no longer the case for exporting, since the ExportedProgram object carries all name and shape and dtype info.
+
+Subsequently, for exporting, dynamic shape is no longer specified through the `inputs` arg of `ct.convert`. Instead, user will need to [express dynamism in torch.export](https://pytorch.org/docs/stable/export.html#expressing-dynamism).
```
This needs to be expanded. 
After a user uses the API of `torch.export.Dim` to export a model with dynamic inputs, what do they need to do to convert it to Core ML? 
Do they still need to provide ct.enumerated shapes or ct.rangeDim as pointed on [this page](https://apple.github.io/coremltools/docs-guides/source/flexible-inputs.html), or now that will be automatically inferred ?

aseemw commented on docs-guides/source/convert-llm.md:
```diff
@@ -0,0 +1,153 @@
+# Converting a Large Language Model
```
You dont need to be generic here. Just say converting openELM model. 
This page is about a specific model only , so the title should reflect that.

YifanShenSZ replied:
Re-titled as suggested
```
Converting an Open Efficient Language Model
```

YifanShenSZ commented on docs-guides/source/model-exporting.md:
````diff
@@ -0,0 +1,111 @@
+    single: ExportedProgram
+```
+
+# Model Exporting
````
Added the context to section `Exporting Limitations`

YifanShenSZ commented on docs-guides/source/model-exporting.md:
```diff
@@ -0,0 +1,111 @@
+
+# Model Exporting
+
+The easiest way to generate ExportedProgram for your model is to use PyTorch's [torch.export.export](https://pytorch.org/docs/stable/export.html#torch.export.export). Exporting runs an example input tensor through your model, and captures the operations that are invoked as that input makes its way through the model's layers.
```
Ah I was simply using the same wording as the [model tracing doc](https://apple.github.io/coremltools/docs-guides/source/model-tracing.html). Reworded to "recommended" as suggested

YifanShenSZ commented:
Tried the [GPT-2 example](https://yifanshensz.github.io/coremltools/docs-guides/source/convert-nlp-model.html) with torch.export, failed due to data-dependent control flow. But, actually, that data-dependent control flow does not make much sense: We don't have to make the decoding loop part of the model itself 🤷 Nowadays we always use c++/swift driver code for the loop and to call single model reference in loop body

aseemw commented on docs-guides/source/convert-llm.md:
```diff
@@ -0,0 +1,153 @@
+# Converting an Open Efficient Language Model
```
Could rename the file as well to `convert-open-elm.md`

YifanShenSZ replied:
Renamed as suggested

YifanShenSZ commented:
[Model scripting](https://yifanshensz.github.io/coremltools/docs-guides/source/model-scripting.html) and [Converting a Natural Language Processing Model](https://yifanshensz.github.io/coremltools/docs-guides/source/convert-nlp-model.html) are now hidden, i.e. no longer displayed in table of contents at left.

`torch.jit.script` is still mentioned in [PyTorch Conversion Workflow](https://yifanshensz.github.io/coremltools/docs-guides/source/convert-pytorch-workflow.html) and [Model Tracing](https://yifanshensz.github.io/coremltools/docs-guides/source/model-tracing.html) along side `torch.jit.trace`, as an experimental support for data-dependent control flow

aseemw commented on docs-guides/source/convert-pytorch-workflow.md:
```diff
@@ -33,64 +46,80 @@ out = traced_model(example_input)
-To ensure that operations such as dropout are disabled, it's important to set the model to evaluation mode (_not_ training mode) before tracing. This setting also results in a more optimized version of the model for conversion.
+### ExportedProgram
+
+[ExportedProgram](https://pytorch.org/docs/stable/export.html) is a new ntermediate representation of a PyTorch model introduced in PyTorch 2. To generate an ExportedProgram representation from PyTorch code, use PyTorch's exporter ([`torch.export.export`](https://pytorch.org/docs/stable/export.html#torch.export.export)) to _export_ the model, as shown in the following example:
```
typo : `ntermediate`

YifanShenSZ replied:
Revised as suggested

aseemw commented on docs-guides/source/convert-pytorch-workflow.md:
````diff
@@ -33,64 +46,80 @@ out = traced_model(example_input)
+model_from_export.save("newmodel_from_export.mlpackage")
 ```
 
 As an alternative, you can convert the model to a neural network by eliminating the `convert_to` parameter:
````
I think we can get rid of this section now. We dont want to recommend neuralnetwork anymore (unless the user selects a low deployment target and automatically get an mlpackage with the nn format)

YifanShenSZ replied:
Removed the neural network section as suggested

aseemw commented on docs-guides/source/convert-pytorch-workflow.md:
````diff
@@ -33,64 +46,80 @@ out = traced_model(example_input)
 ```
 
-If your model uses a data-dependent control flow, such as a loop or conditional, the traced model won't generalize to other inputs. In such cases you can experiment with applying PyTorch's JIT script ([`torch.jit.script`](https://pytorch.org/docs/stable/generated/torch.jit.script.html)) to your model as described in [Model Scripting](model-scripting). You can also use a combination of tracing and ...
+Core ML Tools torch.export conversion support is newly added in 8.0, currently in beta state, in line with the export API status in PyTorch. As of Core ML Tools 8.0, representative models such as mobile bert, resnet, vit, [mobile net](convert-a-torchvision-model-from-pytorch), [deep lab](convert-a-pytorch-segmentation-model), [openelm](convert-openelm) can be converted, and total torch op transla ...
````
`We would recommend to use this actively developing path` : this sentence doesn't make sense. 
If its in active development, why is it recommended? 

I think for now we can skip this line. Once we have more tests passing, then we can add a line here, sth to the effect of,   to try this path (as jit.trace is in maintenance mode) and if any issues are encountered (e.g. lower performance compared to jit.trace path), to report it on Github via an issue.

YifanShenSZ replied:
Rephrased to
```
We would recommend to start trying this path, as PyTorch is moving away from torch.jit.trace
```

aseemw commented on docs-guides/source/convert-pytorch-workflow.md:
```diff
@@ -12,9 +12,9 @@
 The Core ML Tools [Unified Conversion API](unified-conversion-api) produces Core ML models for iOS 13, macOS 10.15, watchOS 6, tvOS 13 or newer deployment targets. If your primary deployment target is iOS 12 or earlier, you can find limited conversion support for PyTorch models via the [onnx-coreml](https://github.com/onnx/onnx) package.
```
This page reads very weirdly, i.e. it doesn't set the context at all. In the beginning it should give an overview. 
(I know even in current state it's not good. But now we have a chance to improve it. ) 

Basically instead of starting the page with "Obtain the Original PyTorch Model" and then going to the section of "Generate the PyTorch Graph" , we need to add an intro paragraph right at the beginning that will should mention what this page will cover and how to convert the model from torch. 

So it should start off by saying something like the following: 

"
To export the model from PyTorch to Core ML consists of two steps: 
1. capturing the model graph from the torch nn.module. 
2. Convert to coreml by passing the captured graph to  the ct.convert API 

There are two torch graphs that are supported by the convert API : 
- the graph as obtained by torch.jit.trace. The conversion support for this torch graph, has existed since coremltools 4 and is more mature and stable 
- the graph obtained by torch.export.export. The support for exported program has been recently added to coremltools 8.0. This path is still in active development (as the torch.export API itself is), hence support is not as mature / stable as the first option above. 
"

YifanShenSZ replied:
Rewrite the introduction as suggested

Also moved the "torch.export path is new" paragraph to introduction

aseemw commented on docs-guides/source/convert-pytorch-workflow.md:
````diff
@@ -24,7 +24,20 @@ import torchvision
+To ensure that operations such as dropout are disabled, it's important to set the model to evaluation mode (_not_ training mode) before tracing. This setting also results in a more optimized version of the model for conversion.
+```
+
+## Generate a PyTorch Graph Version
````
"Generate a PyTorch Graph Version" : that's an odd title, what does "version" mean?  
I think "Generate the PyTorch Graph"  would make more sense. 

(I know this is what it currently says, and I actually didn't realize it. Well whatever page you touch, you should question it, if its makes sense or not and look to improve it if it doesn't. Existing doc is not perfect. This is a good example of that. )

YifanShenSZ replied:
Rephrased to `Capture the PyTorch Graph`

aseemw commented on docs-guides/source/model-exporting.md:
```diff
@@ -0,0 +1,113 @@
+## Difference from Tracing
+For tracing, `ct.convert` requires the `inputs` arg from user. This is no longer the case for exporting, since the ExportedProgram object carries all name and shape and dtype info.
+
+Subsequently, for exporting, dynamic shape is no longer specified through the `inputs` arg of `ct.convert`. Instead, user will need to [express dynamism in torch.export](https://pytorch.org/docs/stable/export.html#expressing-dynamism), which will then be automatically converted to Core ML [`RangeDim`](https://apple.github.io/coremltools/source/coremltools.converters.mil.input_types.html#rangedim) ...
```
I would suggest to add an "admonition" cell at the end of [this](https://apple.github.io/coremltools/docs-guides/source/flexible-inputs.html#set-the-range-for-each-dimension) section as well, saying that if the source format is an ExportedProgram, which has been exported with the torch.export.Dim option, then rangeDim can be skipped.

YifanShenSZ replied:
Added admonition as suggested

aseemw commented:
Left a few more comments, looks good to me after those are addressed!

aseemw commented on docs-guides/source/convert-pytorch-workflow.md:
```diff
@@ -12,9 +12,15 @@
+2. Convert the PyTorch model graph to Core ML, via the Core ML Tools [Unified Conversion API](unified-conversion-api).
 
-[TorchScript](https://pytorch.org/docs/stable/jit.html) is an intermediate representation of a PyTorch model. To generate a TorchScript representation from PyTorch code, use PyTorch's JIT tracer ([`torch.jit.trace`](https://pytorch.org/docs/stable/generated/torch.jit.trace.html)) to _trace_ the model, as shown in the following example:
+Core ML Tools has been supporting the conversion from torch.jit.trace graph since Core ML Tools 4, so this the more mature and stable path. The conversion from torch.export graph is newly added in Core ML Tools 8.0, currently in beta state, in line with the export API status in PyTorch. As of Core ML Tools 8.0, representative models such as mobile bert, resnet, vit, [mobile net](convert-a-torchvi ...
```
---> (capital case for the first letters)
"
MobileBert, ResNet, ViT, [MobileNet](convert-a-torchvision-model-from-pytorch), 
[DeepLab](convert-a-pytorch-segmentation-model), [OpenELM](convert-openelm) etc can be converted
"


Why links for some models and no links for mobilebert, resnet, vit? If you are putting links, put them for all the models you mention.

aseemw replied:
"Core ML Tools has been supporting the conversion from torch.jit.trace graph since Core ML Tools 4, so this the more mature and stable path. The conversion from torch.export graph is newly added in Core ML Tools 8.0, currently in beta state, in line with the export API status in PyTorch. As of Core ML Tools 8.0, representative models such as mobile bert, resnet, vit, [mobile net](convert-a-torchvision-model-from-pytorch), [deep lab](convert-a-pytorch-segmentation-model), [openelm](convert-openelm) can be converted, and total torch op translation test coverage is roughly ~60%. We would recommend to start trying the torch.export path, as [PyTorch no longer offers new feature development for torch.jit.trace](https://github.com/pytorch/pytorch/issues/103841#issuecomment-1605017153)." 


---->

"
The conversion from a graph captured via `torch.jit.trace` has been supported for many versions of coremltools, hence it is the stable and the more performant path. For now, this is the recommended way for converting PyTorch models to Core ML.  

The conversion from `torch.export` graph has been newly added to Core ML Tools 8.0. 
It is currently in beta state, in line with the export API status in PyTorch. 
As of Core ML Tools 8.0, representative models such as MobileBert<link>, ResNet<link>, ViT<link>, [MobileNet](convert-a-torchvision-model-from-pytorch), 
[DeepLab](convert-a-pytorch-segmentation-model), [OpenELM](convert-openelm) etc can be converted, and the total Torch op translation test coverage is roughly ~60%. 
You can start trying the torch.export path on your models that are working with torch.jit.trace already, so as to gradually move them to the export path as PyTorch also [moves](https://github.com/pytorch/pytorch/issues/103841) its support and development to that path over a period of time. In case you hit issues (e.g. models converted via export path are slower than the ones converted from jit.trace path), please report them on Github. "

YifanShenSZ replied:
> Why links for some models and no links for mobilebert, resnet, vit? If you are putting links, put them for all the models you mention.

Because the 3 links are to our conversion examples, not to the model definitions.

> Breaking into 2 paragraphs

Broke as suggested

aseemw commented on docs-guides/source/convert-pytorch-workflow.md:
```diff
@@ -12,9 +12,15 @@
-[TorchScript](https://pytorch.org/docs/stable/jit.html) is an intermediate representation of a PyTorch model. To generate a TorchScript representation from PyTorch code, use PyTorch's JIT tracer ([`torch.jit.trace`](https://pytorch.org/docs/stable/generated/torch.jit.trace.html)) to _trace_ the model, as shown in the following example:
+Core ML Tools has been supporting the conversion from torch.jit.trace graph since Core ML Tools 4, so this the more mature and stable path. The conversion from torch.export graph is newly added in Core ML Tools 8.0, currently in beta state, in line with the export API status in PyTorch. As of Core ML Tools 8.0, representative models such as mobile bert, resnet, vit, [mobile net](convert-a-torchvi ...
+
+## Obtain the Original PyTorch Model
```
Before this, add a sentence, sth like:

"Now let's take a look at how to convert to Core ML through an example."

YifanShenSZ replied:
Added as suggested
```
Now let us take a closer look at how to convert from PyTorch to Core ML through an example.
```

aseemw commented on docs-guides/source/convert-pytorch-workflow.md:
````diff
@@ -24,73 +30,74 @@ import torchvision
+
+```{admonition} Set the Model to Evaluation Mode
+
+To ensure that operations such as dropout are disabled, it's important to set the model to evaluation mode (_not_ training mode) before tracing. This setting also results in a more optimized version of the model for conversion.
````
`To ensure that operations such as dropout are disabled` ---> 
`To ensure that training time operations such as dropout, batchnorm with moving average mean etc, are disabled`

aseemw replied:
you can remove "not training mode"

YifanShenSZ replied:
Revised as suggested

aseemw commented on docs-guides/source/convert-pytorch-workflow.md:
````diff
@@ -24,73 +30,74 @@ import torchvision
 ```
 
-If your model uses a data-dependent control flow, such as a loop or conditional, the traced model won't generalize to other inputs. In such cases you can experiment with applying PyTorch's JIT script ([`torch.jit.script`](https://pytorch.org/docs/stable/generated/torch.jit.script.html)) to your model as described in [Model Scripting](model-scripting). You can also use a combination of tracing and ...
+The process of exporting takes an example input and traces its flow through the model. You can trace the model by creating an example image input, as shown in the above code using random data. To understand the reasons for exporting and how to export a PyTorch model, see [Model Exporting](model-exporting).
````
`To understand the reasons for exporting and how to export a PyTorch model, see [Model Exporting](model-exporting)` ---->  remove this line. Dont need it.

YifanShenSZ replied:
Removed as suggested

aseemw commented on docs-guides/source/convert-pytorch-workflow.md:
````diff
@@ -24,73 +30,74 @@ import torchvision
 The process of tracing takes an example input and traces its flow through the model. You can trace the model by creating an example image input, as shown in the above code using random data. To understand the reasons for tracing and how to trace a PyTorch model, see [Model Tracing](model-tracing).
 
-```{admonition} Set the Model to Evaluation Mode
+If your model uses a data-dependent control flow, such as a loop or conditional, the traced model won't generalize to other inputs. In such cases you can experiment with applying PyTorch's JIT script ([`torch.jit.script`](https://pytorch.org/docs/stable/generated/torch.jit.script.html)) to your model as described in [Model Scripting](model-scripting). You can also use a combination of tracing and ...
````
----> 

"

If your model uses a data-dependent control flow, such as a loop or conditional, the traced model won't generalize to other inputs. In such cases you can use JIT script ([`torch.jit.script`](https://pytorch.org/docs/stable/generated/torch.jit.script.html)) to capture the graph accurately.  Core ML Tools offers limited support for models that are obtained from torch script API, so it may or may not work for your model. See the page on  [Model Scripting](model-scripting) to learn more. 
"

YifanShenSZ replied:
Revised as suggested

aseemw commented on docs-guides/source/convert-a-pytorch-segmentation-model.md:
````diff
@@ -140,46 +134,57 @@ class WrappedDeeplabv3Resnet101(nn.Module):
+
     ``` {note}
-    This example includes a name for the output to make it easier to extract from the Core ML model's prediction dictionary. To learn more about input options, see [Flexible Input Shapes](flexible-inputs).
+    For torch.export, the input name is inherited from the torch model
````
What if the user wants to provide a custom name themselves and not inherit the name that's in the Torch model? Is there a way to do that?

YifanShenSZ replied:
Unfortunately that's not supported yet. Added a sentence
```
As of Core ML Tools 8.0, the input name cannot be customized yet
```

aseemw commented on docs-guides/source/convert-a-pytorch-segmentation-model.md:
````diff
@@ -140,46 +134,57 @@ class WrappedDeeplabv3Resnet101(nn.Module):
     
     ```python
-    mlmodel = ct.convert(
-        trace,
````
typo throughput: `mlmode_from_trace` ----> `mlmodel_from_trace`

`mlmode_from_export` ----> `mlmodel_from_export`

YifanShenSZ replied:
Oops 😅 Fixed as suggested

aseemw commented on docs-guides/source/convert-a-pytorch-segmentation-model.md:
````diff
@@ -140,46 +134,57 @@ class WrappedDeeplabv3Resnet101(nn.Module):
     
     ```python
-    mlmodel.save("SegmentationModel_no_metadata.mlpackage")
+    mlmode_from_export.save("SegmentationModel_no_metadata.mlpackage")
````
Add `mlmodel_from_trace.save()` as well?

YifanShenSZ replied:
Added as suggested

aseemw commented on docs-guides/source/convert-a-torchvision-model-from-pytorch.md:
````diff
@@ -122,21 +129,32 @@ Convert the model to an ML program using the Core ML Tools [`convert()`](https:/
 )
 ```
 
+Torch.export path does not support image input yet, so it still uses tensor input
````
---> Torch.export path does not support image input yet (as of coremltools 8.0)

YifanShenSZ replied:
Added as suggested

aseemw commented on docs-guides/source/model-exporting.md:
````diff
@@ -0,0 +1,113 @@
+    mlmodel = ct.convert(exported_model)
+    ```
+
+## Difference from Tracing
````
What do you think about adding a section down below here that list the current known stuff where torch.export support is lacking, when compared to tracing, as of coremltools 8.0? 

So mentioning things like ops that are not supported , lack of support for image types, lack of enumerate shape support, lack of support for setting custom names for model inputs and outputs and dtypes etc.

YifanShenSZ replied:
Added as suggested. Except the ops, since it is mentioned to earlier this page

aseemw commented:
For the examples updated here, segmentation / mobile net / openELM, have you verified that the performance is same between (a) model is jit traced and converted and (b) when model is exported and converted ?

aseemw reviewed (changes requested):
Added a few more comments. Its getting closer to finish line, still a little bit more to go:)

YifanShenSZ commented:
> For the examples updated here, segmentation / mobile net / openELM, have you verified that the performance (coreml model latency) is same between (a) model is jit traced and converted and (b) when model is exported and converted ? (not really required for this PR, but I was just curios).

Quickly verified with 2 models, performance is the same
| model name | torch.jit.trace / ms | torch.export / ms |
|-------------|--------------------|------------------| 
| MobileNetV2 | 0.56021842000000022 | 0.5450709600000001 |
| DeepLabV3 | 91.316040090000001 | 91.594755919999983 |

aseemw commented on docs-guides/source/model-exporting.md:
```diff
@@ -0,0 +1,120 @@
+It is currently in beta state, in line with the export API status in PyTorch.
+As of Core ML Tools 8.0, representative models such as MobileBert, ResNet, ViT, [MobileNet](convert-a-torchvision-model-from-pytorch), [DeepLab](convert-a-pytorch-segmentation-model), [OpenELM](convert-openelm) can be converted, and the total PyTorch op translation test coverage is roughly ~60%. You can start trying the torch.export path on your models that are working with torch.jit.trace alread ...
+
+Also, torch.export has limitations, see [PyTorch elaboration](https://pytorch.org/docs/stable/export.html#limitations-of-torch-export)
```
instead of `[PyTorch elaboration]` , just say "see here[link]"

YifanShenSZ replied:
Revised as suggested

aseemw commented on docs-guides/source/model-exporting.md:
```diff
@@ -0,0 +1,120 @@
+
+Subsequently, for exporting, dynamic shape is no longer specified through the `inputs` arg of `ct.convert`. Instead, user will need to [express dynamism in torch.export](https://pytorch.org/docs/stable/export.html#expressing-dynamism), which will then be automatically converted to Core ML [`RangeDim`](https://apple.github.io/coremltools/source/coremltools.converters.mil.input_types.html#rangedim) ...
+
+As of Core ML Tools 8.0, there are several features that the torch.export conversion path is yet to support, compared to the mature torch.jit.trace path
```
"...path is yet to support, compared to the mature torch.jit.trace path. Such as: "

YifanShenSZ replied:
Revised as suggested

aseemw commented:
Looks good! Just two minor comments to address.

YifanShenSZ commented:
Thanks @aseemw for all the super helpful suggestions! 🙏 Our torch.export guides have come out nicely 🚀 Please confirm if the final comments have been addressed in the [final commit](https://github.com/apple/coremltools/pull/2356/commits/ed27e93390c6cc2b3f3ed54de386fd282cd21235)

Also, here is the corresponding [html update PR](https://github.com/apple/coremltools/pull/2357) 🙏 Please take a quick look and see if the html also looks good, thx
